Objects/codeobject.c

```diff
@@ -151,169 +163,343 @@ validate_and_copy_tuple(PyObject *tup)
         return NULL;
     }
-    if (intern_strings(cellvars) < 0) {
+    if (intern_strings(con->freevars) < 0) {
         return NULL;
     }
-    if (intern_string_constants(consts, NULL) < 0) {
+    if (intern_strings(con->cellvars) < 0) {
         return NULL;
     }
 
-    /* Make sure that code is indexable with an int, this is
-       a long running assumption in ceval.c and many parts of
-       the interpreter. */
-    if (PyBytes_GET_SIZE(code) > INT_MAX) {
-        PyErr_SetString(PyExc_OverflowError, "co_code larger than INT_MAX");
-        return NULL;
-    }
+    Py_ssize_t nlocals = PyTuple_GET_SIZE(con->varnames);
+    Py_ssize_t totalargs = (Py_ssize_t)con->argcount +
+                           (Py_ssize_t)con->kwonlyargcount +
+                           ((con->flags & CO_VARARGS) != 0) +
+                           ((con->flags & CO_VARKEYWORDS) != 0);
+    Py_ssize_t ncellvars = PyTuple_GET_SIZE(con->cellvars);
+    Py_ssize_t nfreevars = PyTuple_GET_SIZE(con->freevars);
+    Py_ssize_t nfastlocals = nlocals + ncellvars + nfreevars;
 
     /* Check for any inner or outer closure references */
-    n_cellvars = PyTuple_GET_SIZE(cellvars);
-    if (!n_cellvars && !PyTuple_GET_SIZE(freevars)) {
-        flags |= CO_NOFREE;
+    if (!ncellvars && !nfreevars) {
+        con->flags |= CO_NOFREE;
     } else {
-        flags &= ~CO_NOFREE;
+        con->flags &= ~CO_NOFREE;
     }
 
-    n_varnames = PyTuple_GET_SIZE(varnames);
-    if (argcount <= n_varnames && kwonlyargcount <= n_varnames) {
-        /* Never overflows. */
-        total_args = (Py_ssize_t)argcount + (Py_ssize_t)kwonlyargcount +
-                      ((flags & CO_VARARGS) != 0) + ((flags & CO_VARKEYWORDS) != 0);
-    }
-    else {
-        total_args = n_varnames + 1;
-    }
-    if (total_args > n_varnames) {
-        PyErr_SetString(PyExc_ValueError, "code: varnames is too small");
+    PyCodeObject *co = new_code();
+    if (co == NULL) {
         return NULL;
     }
 
+    Py_INCREF(con->filename);
+    co->co_filename = con->filename;
+    Py_INCREF(con->name);
+    co->co_name = con->name;
+    co->co_flags = con->flags;
+
+    Py_INCREF(con->code);
+    co->co_code = con->code;
+    co->co_firstlineno = con->firstlineno;
+    Py_INCREF(con->linetable);
+    co->co_linetable = con->linetable;
+
+    Py_INCREF(con->consts);
+    co->co_consts = con->consts;
+    Py_INCREF(con->names);
+    co->co_names = con->names;
+
+    Py_INCREF(con->varnames);
+    co->co_varnames = con->varnames;
+    Py_INCREF(con->cellvars);
+    co->co_cellvars = con->cellvars;
+    Py_INCREF(con->freevars);
+    co->co_freevars = con->freevars;
+
+    co->co_argcount = con->argcount;
+    co->co_posonlyargcount = con->posonlyargcount;
+    co->co_kwonlyargcount = con->kwonlyargcount;
+
+    co->co_stacksize = con->stacksize;
+
+    Py_INCREF(con->exceptiontable);
+    co->co_exceptiontable = con->exceptiontable;
+
+    co->co_nfastlocals = nfastlocals;
```

Comment:
Does making the various count fields `Py_ssize_t` increase the size of a code object much? If yes, you would have to range-check these values before assigning them.

Reply:
They should be int.  I'll fix that.